### ISSUE TITLE
Allow for an "ignorePattern" option for the RequireThis check

### ISSUE THREAD

**Issue by mmoayyed** [checkstyle/checkstyle#3077]
Would like to be able to ignore certain fields by this check, such as "logger"

**Comment by romani** [comment]
please provide a reason with link/copy-paste to real code.
Please provide details/examples as much as possible.

**Comment by mmoayyed** [comment]
I would want an option to ignore such fields:

```
private final Logger logger = ...
```

In my project, we don't need this access for logger fields.

**Comment by romani** [comment]
please put in web whole class as example. I still do not see why you want to bother yourself with "this." for all except for certain fields, even `logger`.
Please provide more rationale for your idea.

**Comment by mmoayyed** [comment]
Here is an example:
https://github.com/Jasig/cas/blob/master/cas-server-core-events/src/main/java/org/jasig/cas/support/events/CasProxyTicketGrantedEvent.java

I don't know if I can tell you much more than that. It's a matter of project policy, style and some cases personal taste. We typically use this for all instance fields, but not for logger fields. "logger" is an idiom we use. That's it. 

If you are of the opinion that using this is unnecessary, perhaps the check should be removed. If you do offer the check though, it would be nice to control it to suite various styles.

**Comment by romani** [comment]
> It's a matter of project policy, style and some cases personal taste.

yes I know, that is why I asking for as much details as possible. Extending Checkstyle with inaccurate properties is a problem in future as removing it or any other changes will cause braking compatibility, some of them hard to resolve (we have a lot of plugins, extensions). It is pain for our users.

> If you are of the opinion that using this is unnecessary, perhaps the check should be removed.

We can not remove Check, it will damage all config s in whole world, nightmare. We know that users use it, you can find this Check usage on github easily.
We did special update for a Check recently to make it reasonable - http://checkstyle.sourceforge.net/config_coding.html#RequireThis , see property "validateOnlyOverlapping". In this mode Check become less aggressive and old fashioned and this is how we use it - https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L231
new option appear in http://checkstyle.sourceforge.net/releasenotes.html#Release_6.17 .
If you still do not see it matching your idea of style please deactivate the Check it is completely ok.

>  it would be nice to control it to suite various styles. 

I will not close this issue, I will think about it for some time. Most likely you you will be ok with some Filter for certain Check that ignore violations base on message pattern.
I just imaging.... :
```
<module name="SuppressByMessageFilter">
  <property name="checkFormat" value="RequireThis"/>
  <property name="messageFormat" value="^Reference to instance variable 'logger' needs \".*this.\".$"/>
</module>
```
will it work for you ?

**Comment by mmoayyed** [comment]
Your suppression filter suggestion would be OK; What I don't like about it is that it ties the message format in the suppression to the message configured inside the CS rule. That is duplication I would like to not have. What I mostly had in mind was something along the lines of:

```xml
<module name="RequireThis">
...
  <property name="ignorePattern" value="logger|log|WhateverElse"/>
</module>
```

Which would allow one to say: require `this` for everything, except the following pattern matches. That way, I could do away with the coupling and duplication of the error message. 

It would also be just as well if a suppression filter could operate and ignore the check, not based on the reported message, but based on the faulty line in question. Something like:

```
<module name="SuppressByCodeFilter">
  <property name="checkFormat" value="RequireThis"/>
  <property name="ignorePattern" value="logger|log"/>
</module>
```

Which would then allow me to have the `RequireThis` filter without any extra props, but suppress it based on the actual rule rather than the message.

**Comment by romani** [comment]
> What I mostly had in mind was something along the lines of

I do understand your intention, we already have this in number of Checks. I do not like this approach it is a hack. In your case it is more vivid. We will not update the Check to have such option. We need to have Filter to let user do whatever they want with cases they want to exclude/skip.

> It would also be just as well if a suppression filter 

yes, this what I meant , by "ignorePattern" user can define exact message to be precise or small chunk of it.
Config will be for your case:
```
<module name="SuppressByMessageFilter">
  <property name="checkFormat" value="RequireThis"/>
  <property name="ignorePattern" value=".*(logger|log).*"/>
</module>
```

**Comment by romani** [comment]
@mmoayyed , thanks for discussion. New issue is opened to address it #3082 .

I am closing this issue as wont fix.
